### DISCUSSION TITLE
zfs holds being set silently

### DISCUSSION THREAD

**Discussion by hput** [psy0rz/zfs_autobackup#127] [Q&A]
I suspect much of the problem I'm seeing is my own doing due to lack of due diligence (reading docs)
BUT,  I've competely trashed two vms.  I mean took them to the curb experimenting with zfs-autobackup

Apparently za scatters zfs holds throughout a pull with nary a word.  When backup vm is rebooted for any reason, swaths of the zfs filesystems are unmountable apparently and causing the vm to be non-bootable.  At first I had no idea what was going on and my own pestilential fiddling broke the two vms irrevocably.
Currently running two openindiana VBox vms whereon I have installed pip and zfs-autobackup.
A side issue may be that the openindiana crew have decided to pull away from python and phasing it out.
It has an older version ... something before 2.7.
Even with that I think pip would pull out a nicely uptodate `zfs_autobackup'

Back to the vms.  I see quite a list of `holds'.  Maybe not as many as I though but 2 runs of pull backup on only a fewish ZFS fs
I find 12 holds in there:
I setup the pull so the returning fs would start at backup server p0 pool so p0/ssh-client-server/ and a few more ontop in a few cases.
I can find and remove the holds now that I see what is happening but  isn't there a nifty way to get them all at once especially since zfs_autobackup so  gallantly and librally throw them in?

'

**Comment by psy0rz** [comment]
> Apparently za scatters zfs holds throughout a pull with nary a word. When backup vm is rebooted for any reason, swaths of the zfs filesystems are unmountable apparently and causing the vm to be non-bootable. At first I had no idea what was going on and my own pestilential fiddling broke the two vms irrevocably. Currently running two openindiana VBox vms whereon I have installed pip and zfs-autobackup. A side issue may be that the openindiana crew have decided to pull away from python and phasing it out. It has an older version ... something before 2.7. Even with that I think pip would pull out a nicely uptodate `zfs_autobackup'
> 

Not sure what you mean... zfs-autobackup should not touch the stuff its pulling, except it will indeed add holds. These should not be able to break stuff on your vm's.

Each dataset pulled by zfs-autobackup should normally only have one hold, on the latest common snapshot.


> Back to the vms. I see quite a list of `holds'. Maybe not as many as I though but 2 runs of pull backup on only a fewish ZFS fs I find 12 holds in there: I setup the pull so the returning fs would start at backup server p0 pool so p0/ssh-client-server/ and a few more ontop in a few cases. I can find and remove the holds now that I see what is happening but isn't there a nifty way to get them all at once especially since zfs_autobackup so gallantly and librally throw them in?

Not exactly sure what you mean.

Please provide some output. 

If you want to see all the holds on your system, use: 
```
# zfs list -H -o name -t snapshot|xargs -i -n1 zfs holds -H "{}"
```

**Reply by hput** [reply]
thanks for your response.  Its proving rather difficult to get output since It will not boot nor is ssh an option.  I do have a couple of screen shots that show the failure to mount things.
However so as not to waste to  much of your time, I'd like to try another approach.  I made two VMs and have not used zfs_autobackup on the other one yet.  I think, if you are willing, I'd like to get pip and zfs_autobackup installed and try to do a better job.

Probably less work and more learning if I scrap the one I'm crying about and let you coach me a bit if you are willing.
My aim in all this is First to learn to use zfs_autobackup with something approaching a bit of skill and use it to backup two zfs hosts.
I'd hoped to use za to snapshot on both ssh-server-clients and to backup on one at first.  Till I get the hang of it..

So,  can you tell me how you would structure the backup in general?
What I did was send za out to one remote and have it mark the zfs fs to retrieve.
All hosts of mine have a similar naming scheme regarding the pools  rpool on all then p0 which is my name for a pool where I do any work.

So assuming the backup-host pools are named similarly.  My back up point put in place before taking a pull backup  would be on vm p0/NameOfRemote.  
[ I'd send va over to# zfs list -H -o name -t snapshot|xargs -i -n1 zfs holds -H "{}" ssh-client-server and mark the fs to be backed up with test1]
On-test-backup-server, after pull backup and since the remote name is 2x2, I would see something like `p0/2x2/p0/usr/local'.

Am I already off in the weeds or does that sound at least semi-sensible?

PS: thanks for the # zfs list -H -o name -t snapshot|xargs -i -n1 zfs holds -H "{}", but I had already used:
   "zfs get userrefs", thats how I new there were 12 holds put into place.

I won't be doing anymore on this today.  Thanks for your helpful input

**Reply by hput** [reply]
OOPS your command got tangled in there off clipboard  I meant:
[ I'd send va over to ssh-client-server to mark fs to be backed up] I don't see any way to edit it so just revise here I guess

**Comment by psy0rz** [comment]
did you read the getting started page on the wiki?

**Reply by hput** [reply]
After seeing your comment I made my 8th read of that material, although I don't claim to have understood it all.  I did no more than a quick scan of the Restore section since I'm pretty far from being ready for that.

Did you see something In my comments that make you think I missed something important?  What might that be if so.

**Reply by hput** [reply]
psy0rz commented:
"Not exactly sure what you mean.
Please provide some output."

Managed to salvage the pool from what was the backup-host.  But the rpool didn't make the cut.  And with it the ~/.bash_history file that would show the commands I used.

But I can now at least show the result of zfs_autobackup pull:
I ran a hybrid of the command shown in  `get started':
on  "ssh-source-client" 2x:    zfs set autobackup:t1=true p0/local p0/projects   p1/nbk-root
on "ssh-source-client" 2x2:    zfs set autobackup:t1=true p0/local 
You can see results of  the pull backup in the attached file.  Any comments you might have would be welcome.

zfs-autobackup -v --ssh 2x t1 p0/2x
zfs-autobackup -v --ssh 2x2 t1 p0/2x2

[pullResult.txt](https://github.com/psy0rz/zfs_autobackup/files/8596927/pullResult.txt)
